File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long cardnumber = get_long("Number: ");
    long usedcardnumber = cardnumber;
    int remainder = 0;
    bool evenposition = false;
    int summe = 0;
    int length = 0;
    int multipliedremainder = 0;

    while (usedcardnumber > 0)
    {
        length = length + 1;
        remainder = usedcardnumber % 10;

        if (evenposition)
        {
            multipliedremainder = remainder * 2;

            if (multipliedremainder > 9)
            {
                int remainder1 = multipliedremainder / 10;
                int remainder2 = multipliedremainder % 10;
                summe = summe + remainder1;
                summe = summe + remainder2;
            }
            else
            {
                summe = summe + multipliedremainder;
            }
        }
        else
        {
            summe = summe + remainder;
        }

        evenposition = !evenposition;
        usedcardnumber = (usedcardnumber - remainder) / 10;
    }

    if (summe % 10 == 0)
    {
        // AMEX numbers
        if (length == 15)
        {
            int atest = cardnumber / pow(10, 13);

            if (atest == 34 || atest == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }

        // MASTERCARD numbers or VISA numbers
        else if (length == 16)
        {
            int mtest = cardnumber / pow(10, 14);
            int vtest = cardnumber / pow(10, 15);

            if (mtest == 51 || mtest == 52 || mtest == 53 || mtest == 54 || mtest == 55)
            {
                printf("MASTERCARD\n");
            }
            else if (vtest == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }

        // VISA numbers
        else if (length == 13)
        {
            int vtest = cardnumber / pow(10, 12);
            if (vtest == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week6/sentimental-credit/credit.py
def main():
    try:
        cardnumber = int(input("Cardnumber: "))
    except ValueError:
        print("Your Cardnumber can only contain numbers")
        exit()

    used_cardnumber = cardnumber
    remainder = 0
    multiplied_remainder = 0
    even_position = False
    sum = 0
    length = 0

    while used_cardnumber > 0:
        length += 1
        remainder = used_cardnumber % 10

        if even_position:
            multiplied_remainder = remainder * 2

            if multiplied_remainder > 9:
                remainder1 = multiplied_remainder // 10
                remainder2 = multiplied_remainder % 10
                sum += remainder1
                sum += remainder2
            
            else:
                sum += multiplied_remainder
        
        else:
            sum += remainder
        
        even_position = not even_position
        used_cardnumber = used_cardnumber // 10

    if sum % 10 == 0:
        is_AmericanExpress = length == 15 and (cardnumber // (10**13) == 34 or  cardnumber // (10**13) == 37)
        is_MasterCard = length == 16 and (cardnumber // (10**14) == 51 or  cardnumber // (10**14) == 52 or cardnumber // (10**14) == 53 or cardnumber // (10**14) == 54 or cardnumber // (10**14) == 55)
        is_VISA = (length == 16 and (cardnumber // (10**15) == 4)) or (length == 13 and (cardnumber // (10**12) == 4))

        if is_AmericanExpress:      
             print("AMEX")
            
        elif is_MasterCard:
            print("MASTERCARD")
            
        elif is_VISA:
            print("VISA")

        else:
            print("INVALID\n")
        
    else:
        print("INVALID\n")

if __name__ == "__main__":    
    main() 
